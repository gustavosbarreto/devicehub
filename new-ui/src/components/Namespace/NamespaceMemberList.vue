<template>
  <v-table class="bg-v-theme-surface">
    <thead>
      <tr>
        <th
          v-for="(head, i) in headers"
          :key="i"
          :class="head.align ? `text-${head.align}` : 'text-center'"
        >
          <span
            v-if="head.sortable"
            @click="$emit('clickSortableIcon', head.value)"
            @keypress.enter="$emit('clickSortableIcon', head.value)"
            tabindex="0"
            class="hover"
          >
            {{ head.text }}
            <v-tooltip activator="parent" anchor="top"
              >Sort by {{ head.text }}</v-tooltip
            >
          </span>
          <span v-else> {{ head.text }}</span>
        </th>
      </tr>
    </thead>
    <tbody v-if="members">
      <slot name="rows">
        <tr v-for="(member, i) in members" :key="i">
          <td>
            <v-icon> mdi-account </v-icon>
            {{ member.username }}
          </td>

          <td class="text-center">
            {{ member.role }}
          </td>

          <td class="text-end">
            <v-menu
              location="bottom"
              scrim
              eager
              v-if="!isNamespaceOwner(member.role)"
            >
              <template v-slot:activator="{ props }">
                <v-chip v-bind="props" density="comfortable" size="small">
                  <v-icon>mdi-dots-horizontal</v-icon>
                </v-chip>
              </template>
              <v-list class="bg-v-theme-surface" lines="two" density="compact">
                <v-tooltip
                  location="bottom"
                  :disabled="hasAuthorizationRemoveMember()"
                >
                  <template v-slot:activator="{ props }">
                    <NamespaceMemberEdit
                      :member="member"
                      :v-bind="props"
                      @update="refresh"
                      :notHasAuthorization="!hasAuthorizationEditMember()"
                    />
                  </template>
                  <span> You don't have this kind of authorization. </span>
                </v-tooltip>

                <v-tooltip
                  location="bottom"
                  :disabled="hasAuthorizationRemoveMember()"
                >
                  <template v-slot:activator="{ props }">
                    <NamespaceMemberDelete
                      :v-bind="props"
                      :member="member"
                      @update="refresh"
                      :hasAuthorization="hasAuthorizationRemoveMember()"
                    />
                  </template>
                  <span> You don't have this kind of authorization. </span>
                </v-tooltip>
              </v-list>
            </v-menu>
          </td>
        </tr>
      </slot>
    </tbody>
    <div v-else class="pa-4 text-subtitle-2">
      <p>No data available</p>
    </div>
  </v-table>
</template>

<script lang="ts">
import { defineComponent, ref, onMounted, computed } from "vue";
import { useStore } from "../../store";
import hasPermission from "../../utils/permission";
import { actions, authorizer } from "../../authorizer";
import NamespaceMemberDelete from "./NamespaceMemberDelete.vue";
import NamespaceMemberEdit from "./NamespaceMemberEdit.vue";
import { INotificationsError } from "../../interfaces/INotifications";

export default defineComponent({
  props: {
    namespace: {
      type: Object,
      required: true,
    },
  },
  setup(props) {
    const store = useStore();
    const tenant = computed(() => store.getters["auth/tenant"]);
    const members = computed(() => props.namespace.members);
    const hasAuthorizationEditMember = () => {
      const role = store.getters["auth/role"];
      if (role !== "") {
        return hasPermission(
          authorizer.role[role],
          actions.namespace["editMember"]
        );
      }
      return false;
    };
    const hasAuthorizationRemoveMember = () => {
      const role = store.getters["auth/role"];
      if (role !== "") {
        return hasPermission(
          authorizer.role[role],
          actions.namespace["removeMember"]
        );
      }
      return false;
    };
    const getNamespace = async () => {
      try {
        await store.dispatch("namespaces/get", tenant.value);
      } catch (error: any) {
        if (error.response.status === 403) {
          store.dispatch("snackbar/showSnackbarErrorAssociation");
        } else {
          store.dispatch(
            "snackbar/showSnackbarErrorAction",
            INotificationsError.namespaceLoad
          );
        }
      }
    };
    const refresh = () => {
      getNamespace();
    };

    const isNamespaceOwner = (role: string) => {
      return role === "owner";
    };

    return {
      headers: [
        {
          text: "Username",
          value: "username",
          align: "start",
          sortable: false,
        },
        {
          text: "Role",
          value: "role",
          align: "center",
          sortable: false,
        },
        {
          text: "Actions",
          value: "actions",
          align: "end",
          sortable: false,
        },
      ],
      tenant,
      members,
      hasAuthorizationEditMember,
      hasAuthorizationRemoveMember,
      isNamespaceOwner,
      refresh,
    };
  },
  components: { NamespaceMemberDelete, NamespaceMemberEdit },
});
</script>
