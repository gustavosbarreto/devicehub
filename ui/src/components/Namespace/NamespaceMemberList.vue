<template>
  <v-table class="bg-v-theme-surface">
    <thead>
      <tr>
        <th
          v-for="(head, i) in headers"
          :key="i"
          :class="head.align ? `text-${head.align}` : 'text-center'"
        >
          <span
            v-if="head.sortable"
            @click="$emit('clickSortableIcon', head.value)"
            @keypress.enter="$emit('clickSortableIcon', head.value)"
            tabindex="0"
            class="hover"
          >
            {{ head.text }}
            <v-tooltip
              activator="parent"
              anchor="top"
            >Sort by {{ head.text }}</v-tooltip
            >
          </span>
          <span v-else> {{ head.text }}</span>
        </th>
      </tr>
    </thead>
    <tbody v-if="members">
      <slot name="rows">
        <tr v-for="(member, i) in members" :key="i">
          <td>
            <v-icon> mdi-account </v-icon>
            {{ member.email }}
          </td>

          <td class="text-center">
            {{ member.role }}
          </td>

          <td class="text-center">
            {{ member.status }}
            <v-tooltip
              v-if="member.added_at !== '0001-01-01T00:00:00Z'"
              activator="parent"
              location="bottom"
            >This member was added on {{ formatDate(member.added_at) }}</v-tooltip>
          </td>

          <td class="text-end">
            <v-menu
              location="bottom"
              scrim
              eager
              v-if="!isNamespaceOwner(member.role)"
            >
              <template v-slot:activator="{ props }">
                <v-chip v-bind="props" density="comfortable" size="small">
                  <v-icon>mdi-dots-horizontal</v-icon>
                </v-chip>
              </template>
              <v-list class="bg-v-theme-surface" lines="two" density="compact">
                <v-tooltip
                  location="bottom"
                  class="text-center"
                  :disabled="hasAuthorizationEditMember()"
                >
                  <template v-slot:activator="{ props }">
                    <div :v-bind="props">
                      <NamespaceMemberEdit
                        :member="member"
                        @update="refresh"
                        :notHasAuthorization="!hasAuthorizationEditMember()"
                      />
                    </div>
                  </template>
                  <span> You don't have this kind of authorization. </span>
                </v-tooltip>

                <v-tooltip
                  location="bottom"
                  class="text-center"
                  :disabled="hasAuthorizationRemoveMember()"
                >
                  <template v-slot:activator="{ props }">
                    <div :v-bind="props">
                      <NamespaceMemberDelete
                        :member="member"
                        @update="refresh"
                        :hasAuthorization="hasAuthorizationRemoveMember()"
                      />
                    </div>
                  </template>
                  <span> You don't have this kind of authorization. </span>
                </v-tooltip>
              </v-list>
            </v-menu>
          </td>
        </tr>
      </slot>
    </tbody>
    <div v-else class="pa-4 text-subtitle-2">
      <p>No data available</p>
    </div>
  </v-table>
</template>

<script setup lang="ts">
import { computed } from "vue";
import axios, { AxiosError } from "axios";
import { formatDate } from "../../utils/formateDate";
import { useStore } from "../../store";
import hasPermission from "../../utils/permission";
import { actions, authorizer } from "../../authorizer";
import NamespaceMemberDelete from "./NamespaceMemberDelete.vue";
import NamespaceMemberEdit from "./NamespaceMemberEdit.vue";
import { INotificationsError } from "../../interfaces/INotifications";
import handleError from "@/utils/handleError";

const props = defineProps({
  namespace: {
    type: Object,
    required: true,
  },
});

const headers = [
  {
    text: "Email",
    value: "email",
    align: "start",
    sortable: false,
  },
  {
    text: "Role",
    value: "role",
    align: "center",
    sortable: false,
  },
  {
    text: "Status",
    value: "status",
    align: "center",
    sortable: false,
  },
  {
    text: "Actions",
    value: "actions",
    align: "end",
    sortable: false,
  },
];
const store = useStore();
const tenant = computed(() => store.getters["auth/tenant"]);
const members = computed(() => props.namespace.members);
const hasAuthorizationEditMember = () => {
  const role = store.getters["auth/role"];
  if (role !== "") {
    return hasPermission(
      authorizer.role[role],
      actions.namespace.editMember,
    );
  }
  return false;
};
const hasAuthorizationRemoveMember = () => {
  const role = store.getters["auth/role"];
  if (role !== "") {
    return hasPermission(
      authorizer.role[role],
      actions.namespace.removeMember,
    );
  }
  return false;
};
const getNamespace = async () => {
  try {
    await store.dispatch("namespaces/get", tenant.value);
  } catch (error: unknown) {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      if (axiosError.response?.status === 403) {
        store.dispatch("snackbar/showSnackbarErrorAssociation");
        handleError(error);
      }
    } else {
      store.dispatch(
        "snackbar/showSnackbarErrorAction",
        INotificationsError.namespaceLoad,
      );
      handleError(error);
    }
  }
};
const refresh = () => {
  getNamespace();
};

const isNamespaceOwner = (role: string) => role === "owner";
</script>
