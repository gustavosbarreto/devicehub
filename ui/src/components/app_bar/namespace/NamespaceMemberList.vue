<template>
  <fragment>
    <div class="mt-5">
      <v-row class="text-center mb-2">
        <v-col
          v-for="item in heading"
          :key="item.id"
        >
          <b :data-test="item.title+'-title'">
            {{ item.title }}
          </b>
        </v-col>
      </v-row>

      <v-list class="mb-2">
        <v-list-item
          v-for="item in namespace.members"
          :key="item.id"
        >
          <v-row>
            <v-col cols="1">
              <v-icon>
                mdi-account
              </v-icon>
            </v-col>

            <v-col class="text-start">
              <v-list-item-title :data-test="item.username+'-list'">
                {{ item.username }}
              </v-list-item-title>
            </v-col>

            <v-col
              cols="3"
              class="text-end"
            >
              <v-list-item-title :data-test="item.type+'-list'">
                {{ item.type }}
              </v-list-item-title>
            </v-col>

            <v-spacer />

            <div :data-test="item.username+'-actions-list'">
              <v-col>
                <NamespaceMemberFormDialog
                  :add-user="false"
                  :member="item"
                  data-test="NamespaceMemberFormDialogEdit-component"
                  @update="refresh"
                />

                <NamespaceMemberDelete
                  :member="item"
                  data-test="namespaceMemberDelete-component"
                  @update="refresh"
                />
              </v-col>
            </div>
          </v-row>
        </v-list-item>
      </v-list>
    </div>
  </fragment>
</template>

<script>

import NamespaceMemberDelete from '@/components/app_bar/namespace/NamespaceMemberDelete';
import NamespaceMemberFormDialog from '@/components/app_bar/namespace/NamespaceMemberFormDialog';

export default {
  name: 'NamespaceMemberList',

  components: {
    NamespaceMemberDelete,
    NamespaceMemberFormDialog,
  },

  props: {
    namespace: {
      type: Object,
      required: true,
    },
  },

  data() {
    return {
      heading: [
        {
          id: 'name',
          title: 'Username',
        },
        {
          id: 'role',
          title: 'Role',
        },
        {
          id: 'actions',
          title: 'Actions',
        },
      ],
    };
  },

  computed: {
    tenant() {
      return this.$store.getters['auth/tenant'];
    },
  },

  methods: {
    refresh() {
      this.getNamespace();
    },

    async getNamespace() {
      try {
        await this.$store.dispatch('namespaces/get', this.tenant);
        this.name = this.namespace.name;
      } catch (error) {
        if (error.response.status === 403) {
          this.$store.dispatch('snackbar/showSnackbarErrorAssociation');
        } else {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.namespaceLoad);
        }
      }
    },
  },
};

</script>
