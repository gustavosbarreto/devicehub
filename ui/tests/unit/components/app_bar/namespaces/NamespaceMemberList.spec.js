import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import NamespaceMemberList from '@/components/app_bar/namespace/NamespaceMemberList';

describe('NamespaceMemberList', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  const heading = [
    {
      id: 'name',
      title: 'Username',
    },
    {
      id: 'role',
      title: 'Role',
    },
    {
      id: 'actions',
      title: 'Actions',
    },
  ];

  const namespace = {
    name: 'nsxxx',
    members: [{ username: 'user1', type: 'owner' }, { username: 'user2', type: 'administrator' }, { username: 'user3', type: 'observer' }],
  };

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      tenant: 'xxxx',
    },
    getters: {
      'auth/tenant': (state) => state.tenant,
    },
    actions: {
      'snackbar/showSnackbarErrorAssociation': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
      'namespaces/get': () => {},
    },
  });

  const wrapper = shallowMount(NamespaceMemberList, {
    store,
    localVue,
    stubs: ['fragment'],
    propsData: { namespace },
    vuetify,
  });

  ///////
  // Component Rendering
  ///////

  it('Is a Vue instance', () => {
    expect(wrapper).toBeTruthy();
  });

  it('Renders the component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  ///////
  // Data and Props checking
  ///////

  it('Receive data in props', () => {
    expect(wrapper.vm.namespace).toEqual(namespace);
  });

  it('Compares data with default value', () => {
    expect(wrapper.vm.heading).toEqual(heading);
  });

  it('Proccess data in the computed', () => {
    expect(wrapper.vm.tenant).toEqual('xxxx');
  });

  ///////
  // HTML validation
  ///////

  it('Renders the template with data', () => {
    const { members } = namespace;
    members.forEach((m, i) => {
      expect(wrapper.find(`[data-test="${m.username}-list"]`).text()).toBe(members[i].username);
      expect(wrapper.find(`[data-test="${m.type}-list"]`).text()).toBe(members[i].type);
      expect(wrapper.find(`[data-test="${m.username}-actions-list"]`).exists()).toBe(true);
    });
    heading.forEach((col) => {
      expect(wrapper.find(`[data-test="${col.title}-title"]`).text()).toBe(col.title);
    });
  });
});
