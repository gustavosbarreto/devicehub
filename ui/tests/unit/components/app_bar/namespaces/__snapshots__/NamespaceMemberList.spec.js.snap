// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceMemberList Renders the component 1`] = `
<fragment-stub>
  <div class="mt-5">
    <v-row-stub tag="div" class="text-center mb-2">
      <v-col-stub tag="div"><b data-test="Username-title">
          Username
        </b></v-col-stub>
      <v-col-stub tag="div"><b data-test="Role-title">
          Role
        </b></v-col-stub>
      <v-col-stub tag="div"><b data-test="Actions-title">
          Actions
        </b></v-col-stub>
    </v-row-stub>
    <v-list-stub tag="div" class="mb-2">
      <v-list-item-stub activeclass="" tag="div">
        <v-row-stub tag="div">
          <v-col-stub cols="1" tag="div">
            <v-icon-stub>
              mdi-account
            </v-icon-stub>
          </v-col-stub>
          <v-col-stub tag="div" class="text-start">
            <v-list-item-title-stub tag="div" data-test="user1-list">
              user1
            </v-list-item-title-stub>
          </v-col-stub>
          <v-col-stub cols="3" tag="div" class="text-end">
            <v-list-item-title-stub tag="div" data-test="owner-list">
              owner
            </v-list-item-title-stub>
          </v-col-stub>
          <v-spacer-stub tag="div"></v-spacer-stub>
          <div data-test="user1-actions-list">
            <v-col-stub tag="div">
              <namespacememberformdialog-stub member="[object Object]" data-test="NamespaceMemberFormDialogEdit-component"></namespacememberformdialog-stub>
              <namespacememberdelete-stub member="[object Object]" data-test="namespaceMemberDelete-component"></namespacememberdelete-stub>
            </v-col-stub>
          </div>
        </v-row-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" tag="div">
        <v-row-stub tag="div">
          <v-col-stub cols="1" tag="div">
            <v-icon-stub>
              mdi-account
            </v-icon-stub>
          </v-col-stub>
          <v-col-stub tag="div" class="text-start">
            <v-list-item-title-stub tag="div" data-test="user2-list">
              user2
            </v-list-item-title-stub>
          </v-col-stub>
          <v-col-stub cols="3" tag="div" class="text-end">
            <v-list-item-title-stub tag="div" data-test="administrator-list">
              administrator
            </v-list-item-title-stub>
          </v-col-stub>
          <v-spacer-stub tag="div"></v-spacer-stub>
          <div data-test="user2-actions-list">
            <v-col-stub tag="div">
              <namespacememberformdialog-stub member="[object Object]" data-test="NamespaceMemberFormDialogEdit-component"></namespacememberformdialog-stub>
              <namespacememberdelete-stub member="[object Object]" data-test="namespaceMemberDelete-component"></namespacememberdelete-stub>
            </v-col-stub>
          </div>
        </v-row-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" tag="div">
        <v-row-stub tag="div">
          <v-col-stub cols="1" tag="div">
            <v-icon-stub>
              mdi-account
            </v-icon-stub>
          </v-col-stub>
          <v-col-stub tag="div" class="text-start">
            <v-list-item-title-stub tag="div" data-test="user3-list">
              user3
            </v-list-item-title-stub>
          </v-col-stub>
          <v-col-stub cols="3" tag="div" class="text-end">
            <v-list-item-title-stub tag="div" data-test="observer-list">
              observer
            </v-list-item-title-stub>
          </v-col-stub>
          <v-spacer-stub tag="div"></v-spacer-stub>
          <div data-test="user3-actions-list">
            <v-col-stub tag="div">
              <namespacememberformdialog-stub member="[object Object]" data-test="NamespaceMemberFormDialogEdit-component"></namespacememberformdialog-stub>
              <namespacememberdelete-stub member="[object Object]" data-test="namespaceMemberDelete-component"></namespacememberdelete-stub>
            </v-col-stub>
          </div>
        </v-row-stub>
      </v-list-item-stub>
    </v-list-stub>
  </div>
</fragment-stub>
`;
